From: Erik de Castro Lopo <erikd@mega-nerd.com>
Date: Wed,  7 Oct 2009 20:37:12 +1100
Subject: src_simple() causes a crash with certain inputs
Origin: upstream
Bug-Debian: http://bugs.debian.org/539878

Index: libsamplerate-0.1.7/src/src_sinc.c
===================================================================
--- libsamplerate-0.1.7.orig/src/src_sinc.c
+++ libsamplerate-0.1.7/src/src_sinc.c
@@ -1187,18 +1187,21 @@ prepare_data (SINC_FILTER *filter, SRC_D
 			memmove (filter->buffer, filter->buffer + filter->b_current - half_filter_chan_len,
 							(half_filter_chan_len + len) * sizeof (filter->buffer [0])) ;
 
 			filter->b_current = half_filter_chan_len ;
 			filter->b_end = filter->b_current + len ;
 			} ;
 
 		filter->b_real_end = filter->b_end ;
 		len = half_filter_chan_len + 5 ;
 
+		if (len < 0 || filter->b_end + len > filter->b_len)
+			len = filter->b_len - filter->b_end ;
+
 		memset (filter->buffer + filter->b_end, 0, len * sizeof (filter->buffer [0])) ;
 		filter->b_end += len ;
 		} ;
 
 	return 0 ;
 } /* prepare_data */
 
 
Index: libsamplerate-0.1.7/tests/termination_test.c
===================================================================
--- libsamplerate-0.1.7.orig/tests/termination_test.c
+++ libsamplerate-0.1.7/tests/termination_test.c
@@ -20,20 +20,21 @@
 #include <stdlib.h>
 #include <math.h>
 
 #include <samplerate.h>
 
 #include "util.h"
 
 #define	SHORT_BUFFER_LEN	2048
 #define	LONG_BUFFER_LEN		((1 << 16) - 20)
 
+static void simple_test (int converter) ;
 static void stream_test (int converter, double ratio) ;
 static void init_term_test (int converter, double ratio) ;
 
 static int	next_block_length (int reset) ;
 
 int
 main (void)
 {	static double src_ratios [] =
 	{	0.999900, 1.000100, 0.789012, 1.200000, 0.333333, 3.100000,
 		0.125000, 8.000000, 0.099900, 9.990000, 0.100000, 10.00000
@@ -60,24 +61,52 @@ main (void)
 
 	puts ("\n    Sinc interpolator:") ;
 	for (k = 0 ; k < ARRAY_LEN (src_ratios) ; k++)
 		init_term_test (SRC_SINC_FASTEST, src_ratios [k]) ;
 	puts ("") ;
 	for (k = 0 ; k < ARRAY_LEN (src_ratios) ; k++)
 		stream_test (SRC_SINC_FASTEST, src_ratios [k]) ;
 
 	puts ("") ;
 
+	simple_test (SRC_SINC_FASTEST) ;
+
 	return 0 ;
 } /* main */
 
 static void
+simple_test (int converter)
+{
+	int ilen = 199030, olen = 1000, error ;
+
+	{
+		float in [ilen] ;
+		float out [olen] ;
+		double ratio = (1.0 * olen) / ilen ;
+		SRC_DATA src_data =
+		{	in, out,
+			ilen, olen,
+			0, 0, 0,
+			ratio
+		} ;
+
+		error = src_simple (&src_data, converter, 1) ;
+		if (error)
+		{	printf ("\n\nLine %d : %s\n\n", __LINE__, src_strerror (error)) ;
+			exit (1) ;
+			} ;
+	} ;
+
+    return ;
+} /* simple_test */
+
+static void
 init_term_test (int converter, double src_ratio)
 {	static float input [SHORT_BUFFER_LEN], output [SHORT_BUFFER_LEN] ;
 
 	SRC_DATA	src_data ;
 
 	int k, input_len, output_len, error, terminate ;
 
 	printf ("\tinit_term_test   (SRC ratio = %7.4f) .......... ", src_ratio) ;
 	fflush (stdout) ;
 
